perf(napi/parser): faster fixup of `BigInt`s and `RegExp`s

overlookmotel opened the pull request:
#10791 fixed the performance of `oxc-parser` NPM package by removing the reviver from `JSON.parse` call. However, it replaced it with a complete traversal of the AST on JS side to locate `Literal` nodes and update them.

Instead, generate a list of paths to `Literal` nodes needing fixing on Rust side in `ESTree` serializer.

e.g. for this program:

```js
123n;
foo(/xyz/);
```

the fix paths are:

```json
[
    ["body", 0, "expression"],
    ["body", 1, "expression", "arguments", 2]
]
```

Having the location of nodes which need updating reduces work on JS side, as no need to search the entire AST.

Running `pnpm run bench` (in `napi/parser`) locally shows between 8% and 15% speed-up from this change.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10820?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10820** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10820?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10820?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10819** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10819?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10817** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10817?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10820).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Regression in `estree` benchmark is expected. We're doing more work on Rust side in order to save work on JS side. That benchmark only measures the Rust side.

camc314 commented on napi/parser/wrap.cjs:
```diff
@@ -25,39 +23,32 @@ module.exports.wrap = function wrap(result) {
 function jsonParseAst(programJson) {
-  const program = JSON.parse(programJson);
-  transformNode(program);
+  const { node: program, fixes } = JSON.parse(programJson);
+  for (const fixPath of fixes) {
+    applyFix(program, fixPath);
+  }
   return program;
```
I have no idea as to the perf implications of this... but, couldn't we just do this in the reviver of `JSON.parse`

effectivly
```
function jsonParseAst(programJson) {
   const { node: program, fixes } = JSON.parse(programJson, (key, value) => {
// apply fix in here
});
   return program;
```

this way you only visit the tree once?

(note: i guess it might depend on the runtime's implementation of JSON.parse - the overhead of the native to JS call for the reviver might actually make this slower

overlookmotel replied:
Hahaha! Until earlier today that's exactly what we had. This whole business was prompted by the discovery that using a `JSON.parse` reviver was producing a 3x slow-down. See #10783.

camc314 replied:
aha I see!

Boshen commented:
### Merge activity

* **May 5, 10:33 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(napi/parser): faster fixup of `BigInt`s and `RegExp`s (#10820)

#10791 fixed the performance of `oxc-parser` NPM package by removing the reviver from `JSON.parse` call. However, it replaced it with a complete traversal of the AST on JS side to locate `Literal` nodes and update them.

Instead, generate a list of paths to `Literal` nodes needing fixing on Rust side in `ESTree` serializer.

e.g. for this program:

```js
123n;
foo(/xyz/);
```

the fix paths are:

```json
[
    ["body", 0, "expression"],
    ["body", 1, "expression", "arguments", 2]
]
```

Having the location of nodes which need updating reduces work on JS side, as no need to search the entire AST.

Running `pnpm run bench` (in `napi/parser`) locally shows between 8% and 15% speed-up from this change.
